codec-http2/src/main/java/io/netty/handler/codec/http2/HpackEncoder.java

```diff
@@ -380,142 +402,160 @@ private HeaderEntry getEntryInsensitive(CharSequence name, CharSequence value) {
+                    break;
                 }
-                eldest.remove();
-                size -= eldest.size();
-                return eldest;
+                prev = e;
+                e = e.next;
             }
-            prev = e;
-            e = next;
         }
-        return null;
     }
 
     /**
-     * Remove all entries from the dynamic table.
+     * Returns the bucket of the hash table for the hash code h.
      */
-    private void clear() {
-        Arrays.fill(headerFields, null);
-        head.before = head.after = head;
-        size = 0;
+    private int bucket(int h) {
+        return h & hashMask;
     }
 
-    /**
-     * Returns the index into the hash table for the hash code h.
-     */
-    private int index(int h) {
-        return h & hashMask;
+    private static int hash(int nameHash, int valueHash) {
+        return 31 * nameHash + valueHash;
     }
 
-    /**
-     * A linked hash map HpackHeaderField entry.
-     */
-    private static final class HeaderEntry extends HpackHeaderField {
-        // These fields comprise the doubly linked list used for iteration.
-        HeaderEntry before, after;
+    private static final class NameEntry {
+        private NameEntry next;
 
-        // These fields comprise the chained list for header fields with the same hash.
-        HeaderEntry next;
-        int hash;
+        private final CharSequence name;
+
+        private final int hash;
 
         // This is used to compute the index in the dynamic table.
-        int index;
+        private int counter;
 
-        /**
-         * Creates new entry.
-         */
-        HeaderEntry(int hash, CharSequence name, CharSequence value, int index, HeaderEntry next) {
-            super(name, value);
-            this.index = index;
+        private NameEntry(int hash, CharSequence name, int counter, NameEntry next) {
             this.hash = hash;
+            this.name = name;
+            this.counter = counter;
             this.next = next;
         }
 
-        /**
-         * Removes this entry from the linked list.
-         */
         private void remove() {
-            before.after = after;
-            after.before = before;
-            before = null; // null references to prevent nepotism in generational GC.
-            after = null;
-            next = null;
+            next = null; // null references to prevent nepotism in generational GC.
+        }
+    }
+
+    private static final class NameValueEntry extends HpackHeaderField {
+        // This field comprises the linked list used for implementing the eviction policy.
+        private NameValueEntry after;
+
+        private NameValueEntry next;
+
+        // hash of both name and value
+        private final int hash;
+
+        // This is used to compute the index in the dynamic table.
+        private final int counter;
+
+        private NameValueEntry(int hash, CharSequence name, CharSequence value, int counter, NameValueEntry next) {
```

Comment:
should we use "package-private" if the methods are called from the "outside" ?

Reply:
Sure, we'll use package private methods and fields to avoid compiler generated synthetic methods.